server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java

```diff
@@ -46,49 +46,96 @@
+     */
+    public static final Setting<ByteSizeValue> NODE_DISK_AVAILABLE_BANDWIDTH_SETTING = Setting.byteSizeSetting(
+        "node.disk.allocated_bandwidth",
+        ByteSizeValue.MINUS_ONE,
+        ByteSizeValue.MINUS_ONE,
+        new ByteSizeValue(Long.MAX_VALUE, ByteSizeUnit.BYTES),
+        Property.NodeScope
+    );
+
+    /**
+     * Network's read bandwidth allocated for this node. When both this setting and {@link #NODE_DISK_AVAILABLE_BANDWIDTH_SETTING} are
+     * defined, they are used to adjust the default value for {@link #INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING} per node.
+     */
+    public static final Setting<ByteSizeValue> NODE_NETWORK_AVAILABLE_BANDWIDTH_SETTING = Setting.byteSizeSetting(
+        "node.network.allocated_bandwidth",
+        ByteSizeValue.MINUS_ONE,
+        ByteSizeValue.MINUS_ONE,
+        new ByteSizeValue(Long.MAX_VALUE, ByteSizeUnit.BYTES),
+        Property.NodeScope
+    );
+
+    /**
+     * Scaling factor applied to {@link #INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING} when it is derived from external settings like disk
+     * and network bandwidths. See {@link #recoveryMaxBytesPerSecBasedOnExternalSettings(Settings)}.
+     */
+    public static final Setting<Double> NODE_RECOVERY_MAX_BYTES_PER_SEC_FACTOR_SETTING = Setting.doubleSetting(
+        "node.recovery.max_bytes_per_sec.factor",
+        0.8d,
+        0d,
+        Property.NodeScope
+    );
+
+    static final List<Setting<?>> ALL_MAX_BYTES_PER_SEC_EXTERNAL_SETTINGS = List.of(
+        NODE_DISK_AVAILABLE_BANDWIDTH_SETTING,
+        NODE_NETWORK_AVAILABLE_BANDWIDTH_SETTING
+    );
+
+    static final ByteSizeValue DEFAULT_MAX_BYTES_PER_SEC = new ByteSizeValue(40L, ByteSizeUnit.MB);
+    private static final String INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING_KEY = "indices.recovery.max_bytes_per_sec";
+
+    public static final Setting<ByteSizeValue> INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING = new Setting<>(
+        new Setting.SimpleKey(INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING_KEY),
+        settings -> {
+            final ByteSizeValue value;
+            if (NODE_ROLES_SETTING.get(settings).stream().noneMatch(DiscoveryNodeRole::canContainData)) {
                 // if the node is not a data node, this value doesn't matter, use the default
-                return defaultMaxBytesPerSec.getStringRep();
+                value = DEFAULT_MAX_BYTES_PER_SEC;
+
+            } else if (ALL_MAX_BYTES_PER_SEC_EXTERNAL_SETTINGS.stream().allMatch(setting -> setting.exists(settings))) {
+                // computes max_bytes_per_sec from external information
+                value = recoveryMaxBytesPerSecBasedOnExternalSettings(settings);
+
+            } else {
+                // computes max_bytes_per_sec from memory (for dedicated cold/frozen nodes)
+                value = recoveryMaxBytesPerSecBasedOnMemory(settings);
```

Comment:
I think we should calculate this also in the case of external settings and use the max value of the two. The existing "from memory settings" is our minimum value since we would not want to lower it at least initially.

Reply:
This is a tricky one. I think I'd rather stick with this override behaviour and build the existing from-memory behaviour into the numbers provided by the environment. If we took the max here then we wouldn't be able to properly support slower/cheaper cold nodes.

Reply:
As discussed I updated the code to use the highest limit among the one computed from the available bandwidths information and the one from indices.recovery.max_bytes_per_sec or derived from the node memory. I also added a maximum allowed overcommit to avoid situation where a user configured an too high value.